### ISSUE TITLE
Feature Request: Staircase function

### ISSUE THREAD

**Issue by rdipietro** [tensorflow/tensorflow#2432]
Right now for learning rates we have `exponential_decay`, which is useful but doesn't handle fine-tuned scheduling. For example, regardless of whether we did this manually or with `exponential_decay`, there'd be boilerplate code to use a learning rate of 1.0 for 100000 steps, 0.5 for the next 10000 steps, and 0.1 for all remaining steps.

This could be handled with a staircase function, maybe with usage like this:

```
boundaries = [100000, 110000]
values = [1.0, 0.5, 0.1]
learning_rate = staircase(global_step, boundaries, values)
```

This would create a `Tensor` that evaluates to 1.0 when `x <= 100000`, 0.5 when `x > 100000` and `x <= 110000`, and 0.1 when `x > 110000`.

Shouldn't be hard to build using `tf.case`.

Is this of interest or too specific?

**Comment by rdipietro** [comment]
Update: here is the code without the necessary error checks, conversions, tests that'd be necessary for a pull request. Can do this if we think it might be accepted.

```
def staircase(x, boundaries, values):
    """ Staircase function.
    
    Arguments:
        x: A 0-D Tensor.
        boundaries: A 1-D NumPy array with strictly increasing entries.
        values: A 1-D NumPy array that specifies the values for the intervals
            defined by `boundaries`. (It should therefore have one more entry
            than `boundaries`.)
    
    Returns: A 0-D Tensor. Its value is `values[0]` when `x <= boundaries[0]`,
        `values[1]` when `x > boundaries[0]` and `x <= boundaries[1]`, ..., and
        values[-1] when `x > boundaries[-1]`.
    """
    
    pred_fn_pairs = {}
    pred_fn_pairs[x <= boundaries[0]] = lambda: tf.constant(values[0])
    pred_fn_pairs[x > boundaries[-1]] = lambda: tf.constant(values[-1])
    for lower, upper, value in zip(boundaries[:-1], boundaries[1:], values[1:-1]):
        pred_fn_pairs[(x > lower) & (x <= upper)] = lambda: tf.constant(value)
    
    return tf.case(pred_fn_pairs, lambda: tf.constant(values[0]), exclusive=True)
```

**Comment by mrry** [comment]
This seems like it could be a useful contribution!

**Comment by ppwwyyxx** [comment]
Btw, for very cheap use cases such as adjusting learning rate and other hyper-parameters, what I did is simply to compute the value in python and assign it to the corresponding variable in TF, once a while. This should be easy enough and have negligible overhead.
